### PR TITLE
Return all accounts owned by a program

### PR DESCRIPTION
**Pull request by CriesofCarrots**
Preliminary attempt to extend `bank::get_program_accounts_modified_since_parent()`

Tag @rob-solana

### CODE REVIEW

File: runtime/src/bank.rs [hunk 1]
```diff
@@ -2731,4 +2738,39 @@ mod tests {
...
+        bank1.squash();
+        assert_eq!(bank0.get_program_accounts(&program_id), vec![(pubkey, account.clone())]);
+        assert_eq!(bank1.get_program_accounts(&program_id), vec![(pubkey, account.clone())]);
+
+        let bank2 = Arc::new(new_from_parent(&bank1));
+        let bank3 = Arc::new(new_from_parent(&bank2));
+        bank3.squash();
+        assert_eq!(bank0.get_program_accounts(&program_id), vec![(pubkey, account.clone())]);
+        assert_eq!(bank1.get_program_accounts(&program_id), vec![(pubkey, account.clone())]);
+        assert_eq!(bank2.get_program_accounts(&program_id), vec![]);
```
Comment (rob-solana) [inline runtime/src/bank.rs]: why do you expect bank2 not to have the account stored in the parent?

Comment (CriesofCarrots) [reply runtime/src/bank.rs]: `assert_eq!(bank2.get_program_accounts(&program_id), vec![(pubkey, account.clone())]);` fails

File: runtime/src/accounts.rs [hunk 2]
```diff
@@ -329,6 +329,33 @@ impl Accounts {
...
+        for (fork, _) in ancestors.iter() {
+            let mut single_accumulator: Vec<Vec<(Pubkey, u64, Account)>> = self.accounts_db.scan_account_storage(
+                *fork,
+                |stored_account: &StoredAccount,
+                 _id: AppendVecId,
+                 accum: &mut Vec<(Pubkey, u64, Account)>| {
+                    if stored_account.balance.owner == *program_id {
+                        let val = (
+                            stored_account.meta.pubkey,
+                            stored_account.meta.write_version,
```
Comment (rob-solana) [inline runtime/src/accounts.rs]: this value is may be duplicated across forks, so you need another sort key (fork)

File: runtime/src/accounts.rs [hunk 3]
```diff
@@ -329,6 +329,33 @@ impl Accounts {
...
+                        );
+                        accum.push(val)
+                    }
+                },
+            );
+            accumulator.append(&mut single_accumulator);
+        }
+        let mut versions: Vec<(Pubkey, u64, Account)> =
+            accumulator.into_iter().flat_map(|x| x).collect();
+        versions.sort_by_key(|s| (s.0, (s.1 as i64).neg()));
```
Comment (rob-solana) [inline runtime/src/accounts.rs]: `s.1 as i64` is bad form, I think.  how about `std::u64::MAX - s.1` ?

Comment (CriesofCarrots) [reply runtime/src/accounts.rs]: 👍 This is honestly just copy-pasta from `load_by_program()`

Comment (jackcmay) [reply runtime/src/accounts.rs]: fyi, `cargo clippy -- -W clippy::pedantic` will catch these sorts of things.

File: runtime/src/accounts.rs (hunk 2 above)
Comment (rob-solana) [reply runtime/src/accounts.rs]: actually, nevermind.  AppendVecId is apparently monotonically increasing on a per-accounts_db instance

File: runtime/src/bank.rs (hunk 1 above)
Comment (rob-solana) [reply runtime/src/bank.rs]: yeah, ok.  I don't think scan_account_storage can see roots.  I don't think any facility exists for this, you'll need to write something new at the accounts_db level...

the whole accounts_index thing, where asking for the location of an account in a particular fork will fall back to roots is what's missing from the current scan facility

File: runtime/src/accounts.rs (hunk 2 above)
Comment (CriesofCarrots) [reply runtime/src/accounts.rs]: Ah, good point

**Pull request closed without merging**